File: sdks/nodejs/src/integrations/sendgrid.js
import axios from 'axios';
import { emitAPICallEvent } from '../utilities/sanitizeAPICall.js';

export default class SendGridIntegration {
    constructor(apiKey, options = {}) {
        this.apiKey = apiKey;
        this.options = {
            baseURL: 'https://api.sendgrid.com/v3',
            timeout: 30000,
            ...options
        };
    }

    /**
     * Make an API request with error handling
     * @param {Object} config - Axios request config
     * @returns {Promise<Object>} Response data
     */
    async _request(config) {
        const requestHeaders = { 'Authorization': `Bearer ${this.apiKey}`, 'Content-Type': 'application/json', ...config.headers };
        const startTime = Date.now();

        try {
            const response = await axios({
                ...config,
                headers: requestHeaders,
                timeout: this.options.timeout
            });

            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'sendgrid', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: response.status, statusText: response.statusText },
                duration: Date.now() - startTime, error: null
            });

            return response.data;
        } catch (error) {
            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'sendgrid', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: error.response?.status || 0, statusText: error.response?.statusText || 'Error' },
                duration: Date.now() - startTime, error: error.message
            });
            if (error.response) {
                const status = error.response.status;
                const errors = error.response.data?.errors;
                const errorMessage = errors && errors.length > 0
                    ? errors.map(e => e.message).join('; ')
                    : error.response.statusText;
                throw new Error(`SendGrid API error (${status}): ${errorMessage}`);
            } else if (error.request) {
                throw new Error('SendGrid API error: No response received');
            } else {
                throw new Error(`SendGrid API error: ${error.message}`);
            }
        }
    }

    /**
     * Send an email using the SendGrid Mail Send API
     * @param {Object} params - Email parameters
     * @param {string} params.to - Recipient email address(es), comma-separated for multiple
     * @param {string} params.from - Sender email address
     * @param {string} [params.fromName] - Sender display name
     * @param {string} params.subject - Email subject
     * @param {string} [params.text] - Plain text body
     * @param {string} [params.html] - HTML body
     * @param {string} [params.replyTo] - Reply-to email address
     * @param {Array} [params.cc] - CC email addresses
     * @param {Array} [params.bcc] - BCC email addresses
     * @returns {Promise<Object>} Response with status
     */
    async sendEmail(params) {
        const { to, from, fromName, subject, text, html, replyTo, cc, bcc } = params;

        // Build personalizations
        const toAddresses = (typeof to === 'string' ? to.split(',') : to).map(email => ({ email: email.trim() }));
        const personalization = { to: toAddresses };

        if (cc && cc.length > 0) {
            personalization.cc = (typeof cc === 'string' ? cc.split(',') : cc).map(email => ({ email: email.trim() }));
        }
        if (bcc && bcc.length > 0) {
            personalization.bcc = (typeof bcc === 'string' ? bcc.split(',') : bcc).map(email => ({ email: email.trim() }));
        }

        // Build content
        const content = [];
        if (text) {
            content.push({ type: 'text/plain', value: text });
        }
        if (html) {
            content.push({ type: 'text/html', value: html });
        }
        if (content.length === 0) {
            throw new Error('Either text or html content is required');
        }

        // Build request body
        const data = {
            personalizations: [personalization],
            from: fromName ? { email: from, name: fromName } : { email: from },
            subject,
            content
        };

        if (replyTo) {
            data.reply_to = { email: replyTo };
        }

        return await this._request({
            method: 'POST',
            url: `${this.options.baseURL}/mail/send`,
            data
        });
    }
}

File: sdks/python/src/integrations/sendgrid.py
"""
SendGrid Integration for the zv1 engine.

Provides email sending via the SendGrid Mail Send API.
"""

from __future__ import annotations

import logging
import time
from typing import Any

from src.utilities.sanitize_api_call import emit_api_call_event

try:
    import httpx

    HAS_HTTPX = True
except ImportError:
    HAS_HTTPX = False

logger = logging.getLogger(__name__)


class SendgridIntegration:
    """
    Integration with SendGrid's Mail Send API.
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.sendgrid.com/v3",
        timeout: float = 30.0,
    ) -> None:
        if not HAS_HTTPX:
            raise ImportError(
                "httpx package is required for SendGrid integration. "
                "Install with: pip install httpx"
            )

        self.api_key = api_key
        self.base_url = base_url
        self.timeout = timeout

        self.client = httpx.AsyncClient(
            timeout=timeout,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
            },
        )

    async def _request(
        self,
        method: str,
        url: str,
        json_data: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        """Central request method with API call event emission."""
        start_time = int(time.time() * 1000)
        request_headers = dict(self.client.headers)

        try:
            response = await self.client.request(method, url, json=json_data)

            await emit_api_call_event(getattr(self, "_engine_config", None), {
                "timestamp": start_time, "integration": "sendgrid", "nodeId": None, "nodeType": None,
                "request": {"method": method, "url": url, "headers": request_headers, "body": json_data},
                "response": {"status": response.status_code, "statusText": response.reason_phrase or ""},
                "duration": int(time.time() * 1000) - start_time, "error": None,
            })

            self._check_response(response)
            # SendGrid mail/send returns 202 with empty body on success
            if response.status_code == 202:
                return {"status_code": 202, "message": "Email accepted for delivery"}
            return response.json() if response.content else None

        except Exception as e:
            if "SendGrid API error" not in str(e):
                await emit_api_call_event(getattr(self, "_engine_config", None), {
                    "timestamp": start_time, "integration": "sendgrid", "nodeId": None, "nodeType": None,
                    "request": {"method": method, "url": url, "headers": request_headers, "body": json_data},
                    "response": {"status": 0, "statusText": "Error"},
                    "duration": int(time.time() * 1000) - start_time, "error": str(e),
                })
            raise

    async def send_email(
        self,
        *,
        to: str | list[str],
        from_email: str,
        subject: str,
        from_name: str | None = None,
        text: str | None = None,
        html: str | None = None,
        reply_to: str | None = None,
        cc: str | list[str] | None = None,
        bcc: str | list[str] | None = None,
    ) -> dict[str, Any] | None:
        """
        Send an email using the SendGrid Mail Send API.

        Args:
            to: Recipient email address(es).
            from_email: Sender email address.
            subject: Email subject line.
            from_name: Sender display name.
            text: Plain text body.
            html: HTML body.
            reply_to: Reply-to email address.
            cc: CC email addresses.
            bcc: BCC email addresses.

        Returns:
            Response dict with status info.
        """
        # Build to addresses
        if isinstance(to, str):
            to_list = [{"email": addr.strip()} for addr in to.split(",")]
        else:
            to_list = [{"email": addr.strip()} for addr in to]

        personalization: dict[str, Any] = {"to": to_list}

        if cc:
            cc_list = cc.split(",") if isinstance(cc, str) else cc
            personalization["cc"] = [{"email": addr.strip()} for addr in cc_list]
        if bcc:
            bcc_list = bcc.split(",") if isinstance(bcc, str) else bcc
            personalization["bcc"] = [{"email": addr.strip()} for addr in bcc_list]

        # Build content
        content = []
        if text:
            content.append({"type": "text/plain", "value": text})
        if html:
            content.append({"type": "text/html", "value": html})
        if not content:
            raise Exception("Either text or html content is required")

        # Build request body
        from_obj: dict[str, str] = {"email": from_email}
        if from_name:
            from_obj["name"] = from_name

        data: dict[str, Any] = {
            "personalizations": [personalization],
            "from": from_obj,
            "subject": subject,
            "content": content,
        }

        if reply_to:
            data["reply_to"] = {"email": reply_to}

        return await self._request("POST", f"{self.base_url}/mail/send", json_data=data)

    def _check_response(self, response: httpx.Response) -> None:
        """Check response for errors and raise if needed."""
        if response.status_code >= 400:
            try:
                error_data = response.json()
                errors = error_data.get("errors", [])
                if errors:
                    error_message = "; ".join(e.get("message", "") for e in errors)
                else:
                    error_message = response.reason_phrase or "Unknown error"
                raise Exception(
                    f"SendGrid API error ({response.status_code}): {error_message}"
                )
            except Exception as e:
                if "SendGrid API error" in str(e):
                    raise
                raise Exception(
                    f"SendGrid API error ({response.status_code}): {response.reason_phrase}"
                )

    async def close(self) -> None:
        """Close the HTTP client."""
        await self.client.aclose()
